### ISSUE TITLE
Can't make Custom tags to work

### ISSUE THREAD

**Issue by giuliopons** [Program-O/Program-O#171]
I'm using version 2.6.3 of your nice program-o.
I've a aiml file which includes also this:

```
<category>
  <pattern>WHAT TIME IS IT</pattern>
  <template><php>include_once 'foo.php'; foo();</php></template>
</category>
```

in my chatbot/addons/custom_tags folder I've the foo.php file:

```
<?php
function foo() {
  $out = date('m/d/Y i:h:sP');
  return "The date and time is now $out";
}
?>
```

1) The first problem is that uploading the AIML file the validation has blocked it (even if a message tells me that the uploader will not validate the uploaded files).
To fix this I've modified the line 172 of the upload.php script to jumpo away the validation (I've just added a **false &&**):

```
      elseif (false && !$xml->schemaValidate('aiml.xsd')) {
```

2) The second problem, now that the AIML file is uploaded, is that if I ask **WHAT TIME IS IT**, the response is **include_once 'foo.php'; foo();**

I've checked that my global_config.php file has:

```
define('USE_CUSTOM_TAGS', true);
```

Can you please help me?
Thank you very much, your work is great and I'd like to help you in the development of the Program-O.

Giulio

**Comment by AlekzNet** [comment]
The first part is related to https://github.com/Program-O/Program-O/issues/164

**Comment by Dave-Morton** [comment]
The second part is simply because the parse_php_tag function was intentionally never finished, mainly because it's simply there as an example of how to write a function for parsing custom AIML tags, but also because parsing PHP to execute is a risky proposition at best. For the time being this won't be changing, but you're welcome to edit the file chatbot/addons/custom_tags/custom_tags.php to add in the missing functionality. You'll find the function parse_php_tag around line 37. I strongly recommend implementing some strong sanity checking to the code should you decide to modify the script to get this to work.

As to the first part, it's on the ToDo list, and I'll be working on it over the next week or two. School and work are taking most of my time right now.

**Comment by giuliopons** [comment]
Ok @Dave-Morton thank you very much, I will try to make it by myself. Yes, I'm aware of security problems.

**Comment by Dave-Morton** [comment]
Something to consider, as well is deciding on where to put your included PHP files. Currently there is nothing in place to define the location of such files. I suggest adding a line in config/global_config.php to set up a constant to handle this. Something like the following should work:

`define('_ADDONS_INCLUDE_PATH_', _ADDONS_PATH_ . $path_separator . 'includes' . $path_separator);`

Just make sure to create the proper folder in the correct location, whatever you decide to do. :)

**Comment by giuliopons** [comment]
I've made out how to make custom tags that are not php, and probably that's a better way to create services as your google tag and wiki.

**Comment by Dave-Morton** [comment]
Actually, I've taken care of this part already for the next release. There will be a new folder ("includes") in the addons folder, and the constant declaration will be in chatbot/addons/custom_tags/custom_tags.php - the new line of code will be slightly different from what I posted previously, but not significantly so:

`define('_ADDONS_INCLUDE_PATH_', _ADDONS_PATH_ . "includes$path_separator");`

There was a bit of cleanup to do in the config files, so I tossed this bit in while I was at it. The next release won't be for a couple of days, though. I still have a few things to do before I can push the files.
